File: components/aggregator/src/aggregator/schemas/agent.py
"""Pydantic v2 models for agent WebSocket messages and NATS agent message schemas."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

# =============================================================================
# Session State
# =============================================================================


class AgentSessionState(str, Enum):
    """Agent session lifecycle state."""

    INITIALIZING = "initializing"
    RUNNING = "running"
    AWAITING_INPUT = "awaiting_input"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    TIMED_OUT = "timed_out"


# =============================================================================
# Client-to-Server (WebSocket inbound) models
# =============================================================================


class AgentWSEnvelope(BaseModel):
    """WebSocket message envelope for agent protocol."""

    type: str = Field(..., description="Message type identifier")
    session_id: str | None = Field(
        default=None, description="Session ID (set after session.created)"
    )
    sequence: int | None = Field(default=None, description="Monotonic sequence counter")
    timestamp: str | None = Field(default=None, description="ISO 8601 timestamp")
    payload: dict[str, Any] = Field(default_factory=dict, description="Message payload")


class EndpointRef(BaseModel):
    """Reference to an endpoint by owner and slug."""

    owner: str = Field(..., description="Endpoint owner username")
    slug: str = Field(..., description="Endpoint slug")


class SessionStartPayload(BaseModel):
    """Payload for session.start message."""

    prompt: str = Field(..., description="User's initial prompt")
    endpoint: EndpointRef = Field(..., description="Target agent endpoint")
    satellite_token: str = Field(..., description="Satellite token for space authentication")
    transaction_token: str | None = Field(default=None, description="Pre-authorized billing token")
    peer_token: str | None = Field(default=None, description="Peer token for NATS channel")
    peer_channel: str | None = Field(default=None, description="Peer channel identifier")
    config: dict[str, Any] | None = Field(default=None, description="Agent configuration")
    messages: list[dict[str, str]] | None = Field(default=None, description="Conversation history")


class UserMessagePayload(BaseModel):
    """Payload for user.message."""

    content: str = Field(..., description="Message content")
    satellite_token: str | None = Field(
        default=None, description="Optional refreshed satellite token"
    )


class UserConfirmPayload(BaseModel):
    """Payload for user.confirm."""

    tool_call_id: str = Field(..., description="ID of the tool call to confirm")
    modifications: str | None = Field(default=None, description="Optional argument modifications")


class UserDenyPayload(BaseModel):
    """Payload for user.deny."""

    tool_call_id: str = Field(..., description="ID of the tool call to deny")
    reason: str | None = Field(default=None, description="Reason for denial")


# =============================================================================
# Server-to-Client (WebSocket outbound) event payload models
# =============================================================================


class SessionCreatedPayload(BaseModel):
    """Payload for session.created event."""

    session_id: str = Field(..., description="Unique session identifier")


class AgentThinkingPayload(BaseModel):
    """Payload for agent.thinking event."""

    content: str = Field(..., description="Reasoning content")
    is_streaming: bool = Field(default=False, description="Whether content is being streamed")


class AgentToolCallPayload(BaseModel):
    """Payload for agent.tool_call event."""

    tool_call_id: str = Field(..., description="Unique tool call identifier")
    tool_name: str = Field(..., description="Tool name")
    arguments: dict[str, Any] = Field(default_factory=dict, description="Tool arguments")
    requires_confirmation: bool = Field(
        default=False, description="Whether user confirmation is needed"
    )
    description: str | None = Field(default=None, description="Human-readable description")


class AgentToolResultPayload(BaseModel):
    """Payload for agent.tool_result event."""

    tool_call_id: str = Field(..., description="Matching tool call ID")
    status: str = Field(..., description="'success' or 'error'")
    result: Any = Field(default=None, description="Tool output")
    error: str | None = Field(default=None, description="Error details")
    duration_ms: int | None = Field(default=None, description="Execution time in ms")


class AgentMessagePayload(BaseModel):
    """Payload for agent.message event."""

    content: str = Field(..., description="Message content")
    is_complete: bool = Field(default=True, description="Whether message is complete")


class AgentTokenPayload(BaseModel):
    """Payload for agent.token event."""

    token: str = Field(..., description="Streaming token")


class AgentStatusPayload(BaseModel):
    """Payload for agent.status event."""

    status: str = Field(..., description="Status identifier")
    detail: str = Field(default="", description="Human-readable detail")
    progress: float | None = Field(default=None, ge=0, le=1, description="Progress 0-1")


class AgentRequestInputPayload(BaseModel):
    """Payload for agent.request_input event."""

    prompt: str = Field(..., description="What the agent is requesting")


class AgentErrorPayload(BaseModel):
    """Payload for agent.error event."""

    code: str = Field(..., description="Error code")
    message: str = Field(..., description="Error message")
    recoverable: bool = Field(default=False, description="Whether the session can continue")


class SessionCompletedPayload(BaseModel):
    """Payload for session.completed event."""

    session_id: str = Field(..., description="Session ID")


class SessionFailedPayload(BaseModel):
    """Payload for session.failed event."""

    error: str = Field(..., description="Error message")
    reason: str = Field(default="unknown", description="Failure reason")

File: sdk/typescript/src/models/agent.ts
/**
 * TypeScript type definitions for agent events, session state, and message payloads.
 */

// =============================================================================
// Session State
// =============================================================================

/**
 * Agent session state machine states.
 */
export type AgentSessionState =
  | 'idle'
  | 'connecting'
  | 'running'
  | 'awaiting_input'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'error';

// =============================================================================
// Agent Events (discriminated union)
// =============================================================================

export interface ThinkingEvent {
  type: 'agent.thinking';
  payload: {
    content: string;
    is_streaming: boolean;
  };
}

export interface ToolCallEvent {
  type: 'agent.tool_call';
  payload: {
    tool_call_id: string;
    tool_name: string;
    arguments: Record<string, unknown>;
    requires_confirmation: boolean;
    description?: string;
  };
}

export interface ToolResultEvent {
  type: 'agent.tool_result';
  payload: {
    tool_call_id: string;
    status: 'success' | 'error';
    result?: unknown;
    error?: string;
    duration_ms?: number;
  };
}

export interface AgentMessageEvent {
  type: 'agent.message';
  payload: {
    content: string;
    is_complete: boolean;
  };
}

export interface TokenEvent {
  type: 'agent.token';
  payload: {
    token: string;
  };
}

export interface StatusEvent {
  type: 'agent.status';
  payload: {
    status: string;
    detail: string;
    progress?: number;
  };
}

export interface RequestInputEvent {
  type: 'agent.request_input';
  payload: {
    prompt: string;
  };
}

export interface SessionCreatedEvent {
  type: 'session.created';
  payload: {
    session_id: string;
  };
}

export interface SessionCompletedEvent {
  type: 'session.completed';
  payload: {
    session_id: string;
  };
}

export interface SessionFailedEvent {
  type: 'session.failed';
  payload: {
    error: string;
    reason: string;
  };
}

export interface AgentErrorEvent {
  type: 'agent.error';
  payload: {
    code: string;
    message: string;
    recoverable: boolean;
  };
}

/**
 * Union type of all possible agent events.
 */
export type AgentEvent =
  | ThinkingEvent
  | ToolCallEvent
  | ToolResultEvent
  | AgentMessageEvent
  | TokenEvent
  | StatusEvent
  | RequestInputEvent
  | SessionCreatedEvent
  | SessionCompletedEvent
  | SessionFailedEvent
  | AgentErrorEvent;

// =============================================================================
// Session Options
// =============================================================================

/**
 * Configuration for agent sessions.
 */
export interface AgentConfig {
  maxTokens?: number;
  temperature?: number;
  systemPrompt?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Message in conversation history.
 */
export interface AgentHistoryMessage {
  role: string;
  content: string;
}

/**
 * Options for starting an agent session.
 */
export interface AgentSessionOptions {
  /** The initial prompt */
  prompt: string;

  /** Endpoint in "owner/slug" format or { owner, slug } object */
  endpoint: string | { owner: string; slug: string };

  /** Optional agent configuration */
  config?: AgentConfig;

  /** Optional conversation history */
  messages?: AgentHistoryMessage[];

  /** Optional AbortSignal for cancellation */
  signal?: AbortSignal;
}
